File: leetcode/LinkedListTests/LinkedList_1836.cs
namespace LinkedListTests;

public class LinkedList_1836
{
    class Solution {
        public ListNode DeleteDuplicatesUnsorted(ListNode head) {
            var valueFrequencyMap = new Dictionary<int, int>();
            var fakeHead = new ListNode();
            fakeHead.next = head;
            while(head is not null) {
                if(!valueFrequencyMap.ContainsKey(head.val)) {
                    valueFrequencyMap[head.val] = 0;
                }
                valueFrequencyMap[head.val]++;
                head = head.next;
            }

            var prev = fakeHead;
            var curr = fakeHead.next;
            while(curr is not null) {
                if(valueFrequencyMap[curr.val] > 1) {
                    prev.next = curr.next;
                    curr.next = null;
                    curr = prev;
                }
                prev = curr;
                curr = curr.next;
            }
            return fakeHead.next;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_1836.py
from typing import Optional
from collections import Counter


class ListNode:
    def __init__(self, x=0):
        self.val = x
        self.next = None


class Solution:
    def deleteDuplicatesUnsorted(self, head: Optional[ListNode]) -> Optional[ListNode]:
        valueCount = Counter()
        node = head

        while node:
            valueCount[node.val] += 1
            node = node.next

        dummyHead = ListNode()
        dummyHead.next = head
        prev = dummyHead
        current = head

        while current:
            if valueCount[current.val] > 1:
                prev.next = current.next
                current.next = None
                current = prev
            prev = current
            current = current.next

        return dummyHead.next
